File: src/render/merge.ts
import { parse, stringify } from 'yaml';

/**
 * Merge new forge.yaml fields into existing config without overwriting user values.
 * Adds new keys with defaults, preserves existing values.
 */
export function mergeForgeYaml(existing: string, updated: string): string {
	const existingDoc = parse(existing) as Record<string, unknown>;
	const updatedDoc = parse(updated) as Record<string, unknown>;

	const merged = deepMergeNewOnly(existingDoc, updatedDoc);
	return stringify(merged, { lineWidth: 120 });
}

/**
 * Deep merge that only adds keys from `source` that don't exist in `target`.
 * Never overwrites existing values.
 */
function deepMergeNewOnly(
	target: Record<string, unknown>,
	source: Record<string, unknown>,
): Record<string, unknown> {
	const result = { ...target };

	for (const key of Object.keys(source)) {
		if (!(key in result)) {
			result[key] = source[key];
		} else if (
			isPlainObject(result[key]) &&
			isPlainObject(source[key])
		) {
			result[key] = deepMergeNewOnly(
				result[key] as Record<string, unknown>,
				source[key] as Record<string, unknown>,
			);
		}
		// If key exists in target and isn't an object merge case, keep target's value
	}

	return result;
}

function isPlainObject(val: unknown): val is Record<string, unknown> {
	return typeof val === 'object' && val !== null && !Array.isArray(val);
}

File: internal/render/merge.go
package render

import (
	"gopkg.in/yaml.v3"
)

// MergeForgeYAML merges new forge.yaml fields into existing config
// without overwriting user values. Only adds keys that don't exist.
func MergeForgeYAML(existing, updated string) (string, error) {
	var existingDoc, updatedDoc map[string]any
	if err := yaml.Unmarshal([]byte(existing), &existingDoc); err != nil {
		return existing, err
	}
	if err := yaml.Unmarshal([]byte(updated), &updatedDoc); err != nil {
		return existing, err
	}

	merged := deepMergeNewOnly(existingDoc, updatedDoc)
	out, err := yaml.Marshal(merged)
	if err != nil {
		return existing, err
	}
	return string(out), nil
}

func deepMergeNewOnly(target, source map[string]any) map[string]any {
	result := make(map[string]any, len(target))
	for k, v := range target {
		result[k] = v
	}
	for k, v := range source {
		existing, exists := result[k]
		if !exists {
			result[k] = v
			continue
		}
		// Both are maps → recurse
		em, eOk := existing.(map[string]any)
		sm, sOk := v.(map[string]any)
		if eOk && sOk {
			result[k] = deepMergeNewOnly(em, sm)
		}
		// Otherwise keep existing value
	}
	return result
}
